File: apps/api/src/integrations/_mocks/harvest.ts
import type { HttpRequest, HttpResponse } from '../_http/client.js';
import { jsonResponse, consumeFailureFor } from './transport.js';
import { getMockStore } from './store.js';

// Mock Harvest endpoints:
//   POST /v2/invoices/{id}/messages   — send the drafted invoice
//   POST /v2/projects                 — create project (kickoff leg)
//   POST /v2/time_entries             — log time
//   GET  /v2/time_entries             — list
//   PATCH /v2/projects/{id}           — archive

export async function handleHarvest(req: HttpRequest, path: string): Promise<HttpResponse> {
  const fail = consumeFailureFor(`harvest:${req.method} ${path}`);
  if (fail) return fail;

  const store = getMockStore();

  // Send invoice
  const sendMatch = path.match(/^\/v2\/invoices\/([^/]+)\/messages$/);
  if (req.method === 'POST' && sendMatch) {
    const id = sendMatch[1] ?? '';
    const inv = store.invoices.get(id);
    if (!inv) return jsonResponse(404, { error: 'invoice not found' });
    inv.status = 'sent';
    return jsonResponse(201, { id, status: 'sent', sent_at: new Date().toISOString() });
  }

  // Create project (kickoff leg)
  if (req.method === 'POST' && path === '/v2/projects') {
    const body = req.body as { name: string; client_id: string; budget_hours: number };
    const id = `harvest_proj_${Date.now()}`;
    store.harvest_projects.set(id, {
      id,
      name: body.name,
      client_id: body.client_id,
      budget_hours: body.budget_hours,
    });
    return jsonResponse(201, { id, ...body });
  }

  // Archive project
  const archiveMatch = path.match(/^\/v2\/projects\/([^/]+)$/);
  if (req.method === 'PATCH' && archiveMatch) {
    const id = archiveMatch[1] ?? '';
    if (!store.harvest_projects.has(id)) return jsonResponse(404, { error: 'project not found' });
    return jsonResponse(200, { id, archived: true });
  }

  // Log time
  if (req.method === 'POST' && path === '/v2/time_entries') {
    const body = req.body as {
      user_id: string;
      project_id: string;
      date: string;
      hours: number;
      notes?: string | null;
    };
    const id = `harvest_te_${Date.now()}`;
    store.time_entries.set(id, {
      id,
      user_id: body.user_id,
      project_id: body.project_id,
      date: body.date,
      hours: body.hours,
      notes: body.notes ?? null,
    });
    return jsonResponse(201, { id, ...body });
  }

  // List time entries
  if (req.method === 'GET' && path === '/v2/time_entries') {
    return jsonResponse(200, {
      time_entries: Array.from(store.time_entries.values()),
    });
  }

  return jsonResponse(404, { error: `harvest: unmocked ${req.method} ${path}` });
}

File: apps/api_py/src/linkbook/integrations/mocks/harvest.py
"""Harvest mock — mirrors apps/api/src/integrations/_mocks/harvest.ts."""

from __future__ import annotations

import re
import time
from datetime import datetime, timezone

from ..http import HttpRequest, HttpResponse
from .store import MockTimeEntry, get_mock_store
from .transport import consume_failure_for, json_response


async def handle(req: HttpRequest, path: str) -> HttpResponse:
    fail = consume_failure_for(f"harvest:{req.method} {path}")
    if fail is not None:
        return fail
    store = get_mock_store()

    # Send invoice
    m = re.match(r"^/v2/invoices/([^/]+)/messages$", path)
    if req.method == "POST" and m:
        inv = store.invoices.get(m.group(1))
        if inv is None:
            return json_response(404, {"error": "invoice not found"})
        inv.status = "sent"
        return json_response(
            201, {"id": m.group(1), "status": "sent", "sent_at": datetime.now(timezone.utc).isoformat()}
        )

    # List clients (used by find_or_create_client to look up an existing
    # Harvest client by name before creating a project).
    if req.method == "GET" and path == "/v2/clients":
        return json_response(
            200,
            {
                "clients": [
                    {"id": c["id"], "name": c["name"], "is_active": True}
                    for c in store.harvest_clients.values()
                ]
            },
        )

    # Create client (called when find_or_create_client doesn't find one
    # by name above).
    if req.method == "POST" and path == "/v2/clients":
        body = req.body or {}
        new_id = int(time.time() * 1000)
        store.harvest_clients[str(new_id)] = {"id": new_id, "name": body.get("name", "")}
        return json_response(201, {"id": new_id, "name": body.get("name", "")})

    # Create project
    if req.method == "POST" and path == "/v2/projects":
        body = req.body or {}
        new_id = f"harvest_proj_{int(time.time() * 1000)}"
        store.harvest_projects[new_id] = {
            "id": new_id,
            "name": body.get("name"),
            "client_id": body.get("client_id"),
            "budget_hours": body.get("budget_hours"),
        }
        return json_response(201, {"id": new_id, **body})

    # Archive
    m = re.match(r"^/v2/projects/([^/]+)$", path)
    if req.method == "PATCH" and m:
        if m.group(1) not in store.harvest_projects:
            return json_response(404, {"error": "project not found"})
        return json_response(200, {"id": m.group(1), "archived": True})

    # Log time
    if req.method == "POST" and path == "/v2/time_entries":
        body = req.body or {}
        new_id = f"harvest_te_{int(time.time() * 1000)}"
        store.time_entries[new_id] = MockTimeEntry(
            id=new_id,
            user_id=body.get("user_id", ""),
            project_id=body.get("project_id", ""),
            date=body.get("date", ""),
            hours=body.get("hours", 0),
            notes=body.get("notes"),
        )
        return json_response(201, {"id": new_id, **body})

    # List time
    if req.method == "GET" and path == "/v2/time_entries":
        return json_response(
            200,
            {
                "time_entries": [
                    {
                        "id": e.id,
                        "user_id": e.user_id,
                        "project_id": e.project_id,
                        "date": e.date,
                        "hours": e.hours,
                        "notes": e.notes,
                    }
                    for e in store.time_entries.values()
                ]
            },
        )

    return json_response(404, {"error": f"harvest: unmocked {req.method} {path}"})
